core/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -638,6 +621,54 @@ trait OptimisticTransactionImpl extends TransactionalWrite
     }
   }
 
+  /**
+   * Ensure all native table features required by `currentMetadata` is supported by
+   * `currentProtocol`. This means these features must either have support added in the same
+   * transaction via a table property or be auto-update capable
+   * ([[FeatureAutomaticallyEnabledByMetadata.automaticallyUpdateProtocolOfExistingTables]] is set
+   * to `true`).
+   */
+  private def assertTableFeaturesAutomaticallySupported(
+      testedProtocol: Protocol,
+      currentMetadata: Metadata): Unit = {
+    // Collect features that (1) are already supported by the current protocol or by a new
+    // protocol that the table will be updated to; (2) will be added to the protocol by this txn
+    // via `delta.feature.xxx` configs.
+    val supportedFeaturesInTxn =
+      TableFeatureProtocolUtils.getSupportedFeaturesFromConfigs(currentMetadata.configuration) ++
+        testedProtocol.implicitlySupportedFeatures ++
+        testedProtocol.readerAndWriterFeatureNames.flatMap(TableFeature.featureNameToFeature)
```

Comment:
Just curious, is there a reason we can't combine these two concepts into just one thing like "supportedReaderAndWriterFeautres"? If I understand correctly one of these two will be empty at all times. It just seems a little burdensome to have to remember to always include both in the code